File: structures/binary_tree/bst/bst.go
package bst

//BST binary search tree переименовать

import (
	"errors"
)

const (
	EQUAL     = 0
	LEFT_MORE = 1
	LEFT_LESS = -1
)

type Comparator[T any] func(T, T) int

type BST[T, K any] struct {
	Key  T
	Data K
	//a > b return 1, a == b return 0, a < b return -1
	ComparatorKeys Comparator[T]
	parent         *BST[T, K]
	left           *BST[T, K]
	right          *BST[T, K]
}

// Вставка элемента в дерево
func (tn *BST[T, K]) Insert(key T, data K) error {

	if tn == nil {

		return errors.New("node is null")

	} else if tn.ComparatorKeys(tn.Key, key) == EQUAL {

		return errors.New("value exists")

	} else if tn.ComparatorKeys(tn.Key, key) == LEFT_MORE {

		if tn.left == nil {
			tn.left = &BST[T, K]{Key: key, Data: data, ComparatorKeys: tn.ComparatorKeys, parent: tn}
			return nil

		}

		return tn.left.Insert(key, data)

	} else if tn.ComparatorKeys(tn.Key, key) == LEFT_LESS {

		if tn.right == nil {

			tn.right = &BST[T, K]{Key: key, Data: data, ComparatorKeys: tn.ComparatorKeys, parent: tn}
			return nil

		}

		return tn.right.Insert(key, data)

	}
	return nil
}

// Рекурсивный обход дерева InOrder (left->root->right)
func (tr *BST[T, K]) InOrderRecursive(visit func(key T)) {
	if tr != nil {
		tr.left.InOrderRecursive(visit)
		visit(tr.Key)
		tr.right.InOrderRecursive(visit)
	}
}

// Рекурсивный обход дерева PreOrder (root->left->right)
func (tr *BST[T, K]) PreOrderRecursive(visit func(key T)) {
	if tr != nil {
		visit(tr.Key)
		tr.left.PreOrderRecursive(visit)
		tr.right.PreOrderRecursive(visit)
	}
}

// Рекурсивный обход дерева PostOrder (left->right->root)
func (tr *BST[T, K]) PostOrderRecursive(visit func(key T)) {
	if tr != nil {
		tr.left.PostOrderRecursive(visit)
		tr.right.PostOrderRecursive(visit)
		visit(tr.Key)
	}
}

// Поиск node с заданным значением
func (tr *BST[T, K]) FindNode(key T, compare func(a, b T) int) *BST[T, K] {
	if tr == nil {
		return nil
	}

	if compare(tr.Key, key) == EQUAL {
		return tr
	} else if compare(tr.Key, key) == LEFT_MORE {
		return tr.left.FindNode(key, compare)
	} else if compare(tr.Key, key) == LEFT_LESS {
		return tr.right.FindNode(key, compare)
	}
	return nil
}

// Удаление node с заданным value
func (tr *BST[T, K]) RemoveNode(key T, compare func(a, b T) int) *BST[T, K] {

	if tr == nil {
		return nil
	}

	if compare(tr.Key, key) == LEFT_MORE {
		tr.left = tr.left.RemoveNode(key, compare)
	} else if compare(tr.Key, key) == LEFT_LESS {
		tr.right = tr.right.RemoveNode(key, compare)
	} else {
		if tr.left == nil && tr.right == nil {
			return nil
		} else if tr.left == nil {
			return tr.right
		} else if tr.right == nil {
			return tr.left
		}
		successor := tr.right.findMinInNode()
		tr.Key = successor.Key
		tr.right = tr.right.deleteMin()
	}
	return tr
}

// Найти node с минимальным значением (для правого поддерева)
func (tr *BST[T, K]) findMinInNode() *BST[T, K] {
	if tr.left == nil {
		return tr
	}
	return tr.left.findMinInNode()
}

// Удалить node с минимальным значением
func (tr *BST[T, K]) deleteMin() *BST[T, K] {
	if tr == nil {
		return nil
	}

	if tr.left == nil {
		return tr.right
	}

	tr.left = tr.left.deleteMin()
	return tr
}

File: structures/binary_tree/csharp/tree/BST.cs
namespace tree;


enum CMP
{
    EQUAL = 0,
    LEFT_MORE = 1,
    LEFT_LESS = -1
}

public class BST<T, K> where T : IComparable<T>
{
    public T Key{ get; set; }

    public K Data{ get; set; }

    public BST<T, K>? Parent{ get; set; }

    public BST<T, K>? Left{ get; set; }

    public BST<T, K>? Right{ get; set; }


    public BST(T key, K data){
        this.Key = key;
        this.Data = data;
    }

   private BST(T key, K data, BST<T,K> parent){
        this.Key = key;
        this.Data = data;
        this.Parent = parent;
    }

    // Вставка элемента в дерево
    public void Insert(T key, K data) {
        if (this.Key.CompareTo(key) == ((int)CMP.LEFT_MORE)){
            if (this.Left is null){
                this.Left = new BST<T, K>(key, data, this);
            } else {
                this.Left.Insert(key, data);
            }
        } else if (this.Key.CompareTo(key) == ((int)CMP.LEFT_LESS)){
            if (this.Right is null){
                this.Right = new BST<T, K>(key, data, this);
            } else {
                this.Right.Insert(key, data);
            }
        }
    }
    
    // Рекурсивный обход дерева InOrder (left->root->right)
    public void InOrderRecursive(BST<T, K>? node, Action<T,K> visit ){
        if (node is not null){
            this.InOrderRecursive(node.Left, visit);
            visit(node.Key, node.Data);
            this.InOrderRecursive(node.Right, visit);
        }
    }

    // Рекурсивный обход дерева PreOrder (root->left->right)
    public void PreOrderRecursive(BST<T, K>? node, Action<T,K> visit ){
        if (node is not null){
            visit(node.Key, node.Data);
            this.PreOrderRecursive(node.Left, visit);
            this.PreOrderRecursive(node.Right, visit);
        }
    }

    // Рекурсивный обход дерева PostOrder (left->right->root)
    public void PostOrderRecursive(BST<T, K>? node, Action<T,K> visit){
        if (node is not null){
            this.PostOrderRecursive(node.Left, visit);
            this.PostOrderRecursive(node.Right, visit);            
            visit(node.Key, node.Data);
        }
    }

    // Удаление node с заданным value
    public BST<T, K>? RemoveNode(BST<T, K>? node, T key){
        if(node is null){
            return null;
        }

        if (node.Key.CompareTo(key) == ((int)CMP.LEFT_MORE)){
            node.Left = node!.Left!.RemoveNode(node.Left, key);
        } else if (node.Key.CompareTo(key) == ((int)CMP.LEFT_LESS)){
            node.Right = node!.Right!.RemoveNode(node.Right, key);
        } else {
            if(node.Left is null && node.Right is null){
                return null;
            } else if (node.Left is null){
                return node.Right;
            } else if (node.Right is null){
                return node.Left;
            }

            var successor = node.Right.FindMinNode();
            node.Key = successor!.Key;
            node.Right = node.Right.DeleteMin();
        }
        return node;
    }

    // Найти node с минимальным значением (для правого поддерева)
    private BST<T, K>? FindMinNode(){
        if (this.Left is null) {
            return this;
        }
        return this.Left.FindMinNode();
    }

    // Удалить node с минимальным значением
    private BST<T, K>? DeleteMin(){
        if (this is null){
            return null;
        }

        if (this.Left is null){
            return this.Right;
        }

        this.Left = this.Left.DeleteMin();
        return this;
    }

    // Поиск node с заданным значением
    public BST<T, K>? FindNode(BST<T, K> node, T key){
        if(node is null){
            return null;
        }
        if(node.Key.CompareTo(key) == ((int)CMP.EQUAL)){
            return node;
        } else if(node.Key.CompareTo(key) == ((int)CMP.LEFT_MORE)){
            return FindNode(node.Left, key);
        } else if(node.Key.CompareTo(key) == ((int)CMP.LEFT_LESS)){
            return FindNode(node.Right, key);
        }

        return null;
    }

}
